File: Iverson.Clients/Go/iverson/group_by.go
package iverson

import (
	"fmt"
	"strings"

	pb "github.com/iverson/clients/go/generated"
)

// GroupByBuilder builds a GroupByRequest using a fluent API.
//
// Unlike QueryBuilder, this builder runs a compound GROUP BY — multiple
// metrics (SUM/AVG/MIN/MAX/COUNT) over the same grouping in a single
// round-trip, with optional WHERE, HAVING, JOIN, ORDER BY, and LIMIT.
//
// Example:
//
//	req, err := iverson.NewGroupBy("LineItem").
//	    Keys("ReturnFlag", "LineStatus").
//	    Sum("Quantity", "SumQty").
//	    Avg("Quantity", "AvgQty").
//	    CountAll("CountOrder").
//	    Having("SumQty", pb.SearchOperator_GREATER_THAN, numberValue(0)).
//	    OrderBy("ReturnFlag").
//	    Build()
type GroupByBuilder struct {
	typeName    string
	keys        []string
	metrics     []*pb.MetricSpec
	orderBy     []*pb.SearchSort
	where       []*pb.SearchClause
	having      []*pb.SearchClause
	joins       []*pb.JoinSpec
	whereLogic  pb.SearchLogic
	havingLogic pb.SearchLogic
	limit       int32
	err         error
}

// NewGroupBy creates a GroupByBuilder for the given entity type name.
func NewGroupBy(typeName string) *GroupByBuilder {
	return &GroupByBuilder{
		typeName:    typeName,
		whereLogic:  pb.SearchLogic_AND,
		havingLogic: pb.SearchLogic_AND,
		limit:       10_000,
	}
}

// ── Group-by keys ────────────────────────────────────────────────────────────

// Key adds a single GROUP BY column.
func (g *GroupByBuilder) Key(field string) *GroupByBuilder {
	g.keys = append(g.keys, field)
	return g
}

// Keys adds multiple GROUP BY columns.
func (g *GroupByBuilder) Keys(fields ...string) *GroupByBuilder {
	g.keys = append(g.keys, fields...)
	return g
}

// ── WHERE filter (applied before grouping) ──────────────────────────────────

// Where adds a WHERE (FILTER) clause.
func (g *GroupByBuilder) Where(field string, op pb.SearchOperator, val *pb.SearchValue) *GroupByBuilder {
	if val == nil {
		g.err = fmt.Errorf("field %q: nil search value for operator %v", field, op)
		return g
	}
	g.where = append(g.where, &pb.SearchClause{
		Property:   field,
		Operator:   op,
		Value:      val,
		ClauseType: pb.SearchClauseType_FILTER,
	})
	return g
}

// Not adds a MUST_NOT WHERE clause (excludes matches before grouping).
func (g *GroupByBuilder) Not(field string, op pb.SearchOperator, val *pb.SearchValue) *GroupByBuilder {
	if val == nil {
		g.err = fmt.Errorf("field %q: nil search value for operator %v", field, op)
		return g
	}
	g.where = append(g.where, &pb.SearchClause{
		Property:   field,
		Operator:   op,
		Value:      val,
		ClauseType: pb.SearchClauseType_MUST_NOT,
	})
	return g
}

// WithLogic sets the logic used to combine top-level WHERE clauses. Default: AND.
func (g *GroupByBuilder) WithLogic(logic pb.SearchLogic) *GroupByBuilder {
	g.whereLogic = logic
	return g
}

// ── HAVING (applied after grouping; references output alias names) ─────────

// Having adds a HAVING clause. alias must match a metric's output alias.
func (g *GroupByBuilder) Having(alias string, op pb.SearchOperator, val *pb.SearchValue) *GroupByBuilder {
	if val == nil {
		g.err = fmt.Errorf("having %q: nil search value for operator %v", alias, op)
		return g
	}
	g.having = append(g.having, &pb.SearchClause{
		Property:   alias,
		Operator:   op,
		Value:      val,
		ClauseType: pb.SearchClauseType_FILTER,
	})
	return g
}

// WithHavingLogic sets the logic combining HAVING clauses. Default: AND.
func (g *GroupByBuilder) WithHavingLogic(logic pb.SearchLogic) *GroupByBuilder {
	g.havingLogic = logic
	return g
}

// ── JOIN ──────────────────────────────────────────────────────────────────

// Join adds a join from this type to rightType on the given fields.
// The join kind defaults to INNER; pass an explicit pb.JoinKind to override.
func (g *GroupByBuilder) Join(leftField, rightType, rightField string, opts ...pb.JoinKind) *GroupByBuilder {
	kind := pb.JoinKind_INNER
	if len(opts) > 0 {
		kind = opts[0]
	}
	g.joins = append(g.joins, &pb.JoinSpec{
		LeftType:   g.typeName,
		RightType:  rightType,
		LeftField:  leftField,
		RightField: rightField,
		Kind:       kind,
	})
	return g
}

// ── Metrics — simple field ──────────────────────────────────────────────────

// Sum adds a SUM metric. Default alias: "{field}_sum".
func (g *GroupByBuilder) Sum(field string, alias ...string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_SUM, field, "", resolveAlias(alias, field, "sum"))
}

// Avg adds an AVG metric. Default alias: "{field}_avg".
func (g *GroupByBuilder) Avg(field string, alias ...string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_AVG, field, "", resolveAlias(alias, field, "avg"))
}

// Min adds a MIN metric. Default alias: "{field}_min".
func (g *GroupByBuilder) Min(field string, alias ...string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_MIN, field, "", resolveAlias(alias, field, "min"))
}

// Max adds a MAX metric. Default alias: "{field}_max".
func (g *GroupByBuilder) Max(field string, alias ...string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_MAX, field, "", resolveAlias(alias, field, "max"))
}

// Count adds a COUNT metric on a specific field. Default alias: "{field}_count".
func (g *GroupByBuilder) Count(field string, alias ...string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_COUNT, field, "", resolveAlias(alias, field, "count"))
}

// CountAll adds a COUNT(*) metric — no field. Default alias: "count".
func (g *GroupByBuilder) CountAll(alias ...string) *GroupByBuilder {
	name := "count"
	if len(alias) > 0 {
		name = alias[0]
	}
	return g.addMetric(pb.AggregationType_COUNT, "", "", name)
}

// ── Metrics — expression (raw SQL) ──────────────────────────────────────────

// SumExpr adds a SUM metric over a raw SQL expression, e.g. "Price * (1 - Discount)".
func (g *GroupByBuilder) SumExpr(expression, alias string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_SUM, "", expression, alias)
}

// AvgExpr adds an AVG metric over a raw SQL expression.
func (g *GroupByBuilder) AvgExpr(expression, alias string) *GroupByBuilder {
	return g.addMetric(pb.AggregationType_AVG, "", expression, alias)
}

// ── Ordering and limit ───────────────────────────────────────────────────────

// OrderBy adds a sort clause (typically referencing a group-by key or metric alias).
// Pass true as the optional descending argument for a descending sort.
func (g *GroupByBuilder) OrderBy(field string, descending ...bool) *GroupByBuilder {
	desc := false
	if len(descending) > 0 {
		desc = descending[0]
	}
	g.orderBy = append(g.orderBy, &pb.SearchSort{Property: field, Descending: desc})
	return g
}

// OrderByDesc adds a descending sort clause.
func (g *GroupByBuilder) OrderByDesc(field string) *GroupByBuilder {
	return g.OrderBy(field, true)
}

// Limit sets the max number of grouped rows returned. Default: 10000.
func (g *GroupByBuilder) Limit(n int32) *GroupByBuilder {
	g.limit = n
	return g
}

// ── Build ─────────────────────────────────────────────────────────────────

// Build constructs the GroupByRequest proto. An optional traceId may be supplied.
func (g *GroupByBuilder) Build(traceId ...string) (*pb.GroupByRequest, error) {
	if g.err != nil {
		return nil, g.err
	}

	aliases := map[string]bool{}
	for _, m := range g.metrics {
		key := strings.ToLower(m.Name)
		if aliases[key] {
			return nil, fmt.Errorf("duplicate metric alias %q", m.Name)
		}
		aliases[key] = true
	}
	keys := map[string]bool{}
	for _, k := range g.keys {
		key := strings.ToLower(k)
		if keys[key] {
			return nil, fmt.Errorf("duplicate key %q", k)
		}
		keys[key] = true
		if aliases[key] {
			return nil, fmt.Errorf("key %q collides with an existing metric alias", k)
		}
		aliases[key] = true
	}
	for _, h := range g.having {
		if !aliases[strings.ToLower(h.Property)] {
			return nil, fmt.Errorf(
				"HAVING references %q, which is neither a metric alias nor a key", h.Property)
		}
	}
	for _, s := range g.orderBy {
		if !aliases[strings.ToLower(s.Property)] {
			return nil, fmt.Errorf(
				"OrderBy references %q, which is neither a metric alias nor a key", s.Property)
		}
	}

	id := ""
	if len(traceId) > 0 {
		id = traceId[0]
	}
	return &pb.GroupByRequest{
		TypeName: g.typeName,
		Query: &pb.SearchQuery{
			Clauses: g.where,
			Logic:   g.whereLogic,
		},
		Keys:    g.keys,
		Metrics: g.metrics,
		Having: &pb.SearchQuery{
			Clauses: g.having,
			Logic:   g.havingLogic,
		},
		OrderBy: g.orderBy,
		Limit:   g.limit,
		Joins:   g.joins,
		TraceId: id,
	}, nil
}

// ── Internal helpers ─────────────────────────────────────────────────────────

func (g *GroupByBuilder) addMetric(aggType pb.AggregationType, field, expression, alias string) *GroupByBuilder {
	g.metrics = append(g.metrics, &pb.MetricSpec{
		Name:       alias,
		Type:       aggType,
		Field:      field,
		Expression: expression,
	})
	return g
}

// resolveAlias returns the first element of alias if present, else "{field}_{suffix}".
func resolveAlias(alias []string, field, suffix string) string {
	if len(alias) > 0 {
		return alias[0]
	}
	return fmt.Sprintf("%s_%s", field, suffix)
}

File: Iverson.Clients/TypeScript/src/group-by.ts
/**
 * Fluent GROUP BY query builder that compiles to a GroupByRequest proto.
 *
 * Usage:
 *   const req = groupBy('LineItem')
 *     .key('orderStatus')
 *     .sum('quantity')
 *     .having('quantity_sum', SearchOperator.GREATER_THAN, 100)
 *     .build();
 */

import {
    AggregationType,
    GroupByRequest,
    JoinKind,
    JoinSpec,
    MetricSpec,
    SearchClause,
    SearchClauseType,
    SearchLogic,
    SearchOperator,
    SearchQuery,
    SearchSort,
} from '../generated/object_search.js';
import { toSearchValue } from './search.js';

// ── GroupByBuilder ────────────────────────────────────────────────────────────

/**
 * Fluent DSL builder that compiles to a GroupByRequest proto.
 */
export class GroupByBuilder {
    private readonly _typeName: string;
    private readonly _keys: string[] = [];
    private readonly _metrics: MetricSpec[] = [];
    private readonly _orderBy: SearchSort[] = [];
    private readonly _where: SearchClause[] = [];
    private readonly _having: SearchClause[] = [];
    private readonly _joins: JoinSpec[] = [];
    private _whereLogic: SearchLogic = SearchLogic.AND;
    private _havingLogic: SearchLogic = SearchLogic.AND;
    private _limit = 10_000;

    constructor(typeName: string) {
        this._typeName = typeName;
    }

    // ── Group-by keys ─────────────────────────────────────────────────────────

    /** Add a single GROUP BY column. */
    key(field: string): this {
        this._keys.push(field);
        return this;
    }

    /** Add multiple GROUP BY columns. */
    keys(...fields: string[]): this {
        this._keys.push(...fields);
        return this;
    }

    // ── Filtering ─────────────────────────────────────────────────────────────

    /** Add a WHERE clause (applied before grouping). */
    where(field: string, op: SearchOperator, value: unknown): this {
        this._where.push({
            property: field,
            operator: op,
            value: toSearchValue(value),
            clauseType: SearchClauseType.FILTER,
        });
        return this;
    }

    /** Add a MUST_NOT WHERE clause (excludes matches before grouping). */
    not(field: string, op: SearchOperator, value: unknown): this {
        this._where.push({
            property: field,
            operator: op,
            value: toSearchValue(value),
            clauseType: SearchClauseType.MUST_NOT,
        });
        return this;
    }

    /** Set the top-level WHERE clause logic (AND / OR). Default: AND. */
    withLogic(logic: SearchLogic): this {
        this._whereLogic = logic;
        return this;
    }

    /** Add a HAVING clause (applied after grouping); `alias` references a metric output alias. */
    having(alias: string, op: SearchOperator, value: unknown): this {
        this._having.push({
            property: alias,
            operator: op,
            value: toSearchValue(value),
            clauseType: SearchClauseType.FILTER,
        });
        return this;
    }

    /** Set the logic combining HAVING clauses. Default: AND. */
    withHavingLogic(logic: SearchLogic): this {
        this._havingLogic = logic;
        return this;
    }

    // ── Joins ─────────────────────────────────────────────────────────────────

    /** Add an inner/left/right join to another registered type. */
    join(leftField: string, rightType: string, rightField: string, kind: JoinKind = JoinKind.INNER): this {
        this._joins.push({
            leftType: this._typeName,
            rightType,
            leftField,
            rightField,
            kind,
        });
        return this;
    }

    // ── Metrics ───────────────────────────────────────────────────────────────

    private _addMetric(type: AggregationType, field: string, alias: string | undefined): this {
        this._metrics.push({
            name: alias ?? `${field}_${AggregationType[type].toLowerCase()}`,
            type,
            field,
            expression: '',
        });
        return this;
    }

    /** SUM metric. Default alias: `{field}_sum`. */
    sum(field: string, alias?: string): this {
        return this._addMetric(AggregationType.SUM, field, alias);
    }

    /** AVG metric. Default alias: `{field}_avg`. */
    avg(field: string, alias?: string): this {
        return this._addMetric(AggregationType.AVG, field, alias);
    }

    /** MIN metric. Default alias: `{field}_min`. */
    min(field: string, alias?: string): this {
        return this._addMetric(AggregationType.MIN, field, alias);
    }

    /** MAX metric. Default alias: `{field}_max`. */
    max(field: string, alias?: string): this {
        return this._addMetric(AggregationType.MAX, field, alias);
    }

    /** COUNT metric on a specific field. Default alias: `{field}_count`. */
    count(field: string, alias?: string): this {
        return this._addMetric(AggregationType.COUNT, field, alias);
    }

    /** COUNT(*) metric — no field. Default alias: `count`. */
    countAll(alias = 'count'): this {
        this._metrics.push({
            name: alias,
            type: AggregationType.COUNT,
            field: '',
            expression: '',
        });
        return this;
    }

    /** SUM over a raw SQL expression. */
    sumExpr(expression: string, alias: string): this {
        this._metrics.push({
            name: alias,
            type: AggregationType.SUM,
            field: '',
            expression,
        });
        return this;
    }

    /** AVG over a raw SQL expression. */
    avgExpr(expression: string, alias: string): this {
        this._metrics.push({
            name: alias,
            type: AggregationType.AVG,
            field: '',
            expression,
        });
        return this;
    }

    // ── Sorting and limiting ──────────────────────────────────────────────────

    /** Add a sort clause (typically referencing a group-by key or metric alias). */
    orderBy(field: string, descending = false): this {
        this._orderBy.push({ property: field, descending });
        return this;
    }

    /** Add a descending sort clause. */
    orderByDesc(field: string): this {
        return this.orderBy(field, true);
    }

    /** Set the max number of grouped rows returned. Default: 10000. */
    limit(n: number): this {
        this._limit = n;
        return this;
    }

    // ── Build ─────────────────────────────────────────────────────────────────

    /** Compile to a GroupByRequest proto message. */
    build(traceId = ''): GroupByRequest {
        const aliases = new Set<string>();
        for (const m of this._metrics) {
            const key = m.name.toLowerCase();
            if (aliases.has(key)) throw new Error(`Duplicate metric alias '${m.name}'.`);
            aliases.add(key);
        }
        const keys = new Set<string>();
        for (const k of this._keys) {
            const key = k.toLowerCase();
            if (keys.has(key)) throw new Error(`Duplicate key '${k}'.`);
            keys.add(key);
            if (aliases.has(key)) throw new Error(`Key '${k}' collides with an existing metric alias.`);
            aliases.add(key);
        }

        for (const h of this._having) {
            if (!aliases.has(h.property.toLowerCase())) {
                throw new Error(
                    `HAVING references '${h.property}', which is neither a metric alias nor a key.`);
            }
        }
        for (const s of this._orderBy) {
            if (!aliases.has(s.property.toLowerCase())) {
                throw new Error(
                    `orderBy references '${s.property}', which is neither a metric alias nor a key.`);
            }
        }

        const query: SearchQuery = {
            clauses: [...this._where],
            logic: this._whereLogic,
            sort: [],
        };
        const having: SearchQuery = {
            clauses: [...this._having],
            logic: this._havingLogic,
            sort: [],
        };
        return {
            typeName: this._typeName,
            query,
            keys: [...this._keys],
            metrics: [...this._metrics],
            having,
            orderBy: [...this._orderBy],
            limit: this._limit,
            joins: [...this._joins],
            traceId,
        };
    }
}

/** Start a fluent GROUP BY query for the given entity type. */
export function groupBy(typeName: string): GroupByBuilder {
    return new GroupByBuilder(typeName);
}
